File: workers/python/worker.py
#!/usr/bin/env python3
"""
Python Worker for Task Queue System

This worker polls SQS for tasks and processes them based on their type.
Supports multiple task types with extensible handler architecture.
"""

import json
import logging
import os
import signal
import sys
import time
from typing import Dict, Any, Callable
import boto3
import psycopg2
from psycopg2.extras import RealDictCursor
import requests

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class TaskWorker:
    """Main worker class that processes tasks from SQS"""
    
    def __init__(self):
        self.running = True
        self.worker_id = f"python-worker-{os.getpid()}"
        
        # Initialize AWS SQS client
        self.sqs = boto3.client(
            'sqs',
            region_name=os.getenv('AWS_REGION', 'us-east-1'),
            aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
            aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
        )
        self.queue_url = os.getenv('AWS_SQS_QUEUE_URL')
        
        # Initialize database connection
        self.db_conn = psycopg2.connect(os.getenv('DATABASE_URL'))
        self.db_conn.autocommit = True
        
        # Task handlers mapping
        self.handlers: Dict[str, Callable] = {
            'email': self.handle_email_task,
            'data': self.handle_data_task,
            'file': self.handle_file_task,
            'api': self.handle_api_task,
            'script': self.handle_script_task,
            'report': self.handle_report_task,
        }
        
        # Setup signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self.shutdown)
        signal.signal(signal.SIGTERM, self.shutdown)
        
        logger.info(f"Worker {self.worker_id} initialized")
    
    def shutdown(self, signum, frame):
        """Handle shutdown signals gracefully"""
        logger.info(f"Worker {self.worker_id} shutting down...")
        self.running = False
        if self.db_conn:
            self.db_conn.close()
        sys.exit(0)
    
    def run(self):
        """Main worker loop"""
        logger.info(f"Worker {self.worker_id} starting...")
        
        while self.running:
            try:
                # Poll SQS for messages
                response = self.sqs.receive_message(
                    QueueUrl=self.queue_url,
                    MaxNumberOfMessages=1,
                    WaitTimeSeconds=20,  # Long polling
                    MessageAttributeNames=['All']
                )
                
                messages = response.get('Messages', [])
                
                for message in messages:
                    self.process_message(message)
                    
            except Exception as e:
                logger.error(f"Error in worker loop: {e}")
                time.sleep(5)  # Wait before retrying
    
    def process_message(self, message: Dict[str, Any]):
        """Process a single message from SQS"""
        try:
            # Parse message body
            body = json.loads(message['Body'])
            task_id = body.get('task_id')
            task_type = body.get('type')
            payload = body.get('payload', {})
            
            logger.info(f"Processing task {task_id} of type {task_type}")
            
            # Update task status to processing
            self.update_task_status(message['MessageId'], 'processing')
            
            # Get handler for task type
            handler = self.handlers.get(task_type)
            if not handler:
                raise ValueError(f"Unknown task type: {task_type}")
            
            # Execute task handler
            result = handler(payload)
            
            # Mark task as completed
            self.complete_task(message['MessageId'], result)
            
            # Delete message from queue
            self.sqs.delete_message(
                QueueUrl=self.queue_url,
                ReceiptHandle=message['ReceiptHandle']
            )
            
            logger.info(f"Task {task_id} completed successfully")
            
        except Exception as e:
            logger.error(f"Error processing message: {e}")
            self.fail_task(message.get('MessageId'), str(e))
    
    def update_task_status(self, message_id: str, status: str):
        """Update task status in database"""
        with self.db_conn.cursor() as cursor:
            cursor.execute("""
                UPDATE tasks 
                SET status = %s, worker_id = %s, started_at = CURRENT_TIMESTAMP
                WHERE message_id = %s
            """, (status, self.worker_id, message_id))
    
    def complete_task(self, message_id: str, result: Dict[str, Any]):
        """Mark task as completed with result"""
        with self.db_conn.cursor() as cursor:
            cursor.execute("""
                UPDATE tasks 
                SET status = 'completed', 
                    result = %s,
                    completed_at = CURRENT_TIMESTAMP
                WHERE message_id = %s
            """, (json.dumps(result), message_id))
    
    def fail_task(self, message_id: str, error: str):
        """Mark task as failed with error message"""
        with self.db_conn.cursor() as cursor:
            cursor.execute("""
                UPDATE tasks 
                SET status = 'failed', 
                    error_message = %s,
                    completed_at = CURRENT_TIMESTAMP
                WHERE message_id = %s
            """, (error, message_id))
    
    # Task handlers
    def handle_email_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle email processing tasks"""
        recipient = payload.get('recipient')
        subject = payload.get('subject')
        body = payload.get('body')
        
        # Simulate email sending
        logger.info(f"Sending email to {recipient}: {subject}")
        time.sleep(2)  # Simulate processing time
        
        return {
            'status': 'sent',
            'recipient': recipient,
            'timestamp': time.time()
        }
    
    def handle_data_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle data manipulation tasks"""
        operation = payload.get('operation')
        data = payload.get('data', [])
        
        # Simulate data processing
        logger.info(f"Processing data operation: {operation}")
        
        if operation == 'sum':
            result = sum(data)
        elif operation == 'average':
            result = sum(data) / len(data) if data else 0
        elif operation == 'count':
            result = len(data)
        else:
            result = data
        
        time.sleep(1)  # Simulate processing time
        
        return {
            'operation': operation,
            'result': result,
            'items_processed': len(data)
        }
    
    def handle_file_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle file operation tasks"""
        operation = payload.get('operation')
        file_path = payload.get('file_path')
        
        logger.info(f"Processing file operation: {operation} on {file_path}")
        
        # Simulate file operations
        time.sleep(3)
        
        return {
            'operation': operation,
            'file_path': file_path,
            'status': 'completed',
            'size_bytes': 1024  # Mock value
        }
    
    def handle_api_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle API call tasks"""
        url = payload.get('url')
        method = payload.get('method', 'GET')
        headers = payload.get('headers', {})
        data = payload.get('data')
        
        logger.info(f"Making API call: {method} {url}")
        
        try:
            response = requests.request(
                method=method,
                url=url,
                headers=headers,
                json=data,
                timeout=30
            )
            
            return {
                'status_code': response.status_code,
                'response': response.json() if response.headers.get('content-type', '').startswith('application/json') else response.text,
                'headers': dict(response.headers)
            }
        except Exception as e:
            return {
                'error': str(e),
                'status': 'failed'
            }
    
    def handle_script_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle script execution tasks"""
        script_name = payload.get('script_name')
        args = payload.get('args', [])
        
        logger.info(f"Executing script: {script_name} with args: {args}")
        
        # Simulate script execution
        time.sleep(5)
        
        return {
            'script': script_name,
            'args': args,
            'exit_code': 0,
            'output': 'Script executed successfully'
        }
    
    def handle_report_task(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Handle report generation tasks"""
        report_type = payload.get('report_type')
        parameters = payload.get('parameters', {})
        
        logger.info(f"Generating report: {report_type}")
        
        # Simulate report generation
        time.sleep(10)
        
        return {
            'report_type': report_type,
            'status': 'generated',
            'file_url': f'/reports/{report_type}_{int(time.time())}.pdf',
            'pages': 42
        }


if __name__ == '__main__':
    # Check required environment variables
    required_env = [
        'AWS_REGION',
        'AWS_ACCESS_KEY_ID',
        'AWS_SECRET_ACCESS_KEY',
        'AWS_SQS_QUEUE_URL',
        'DATABASE_URL'
    ]
    
    missing = [var for var in required_env if not os.getenv(var)]
    if missing:
        logger.error(f"Missing required environment variables: {missing}")
        sys.exit(1)
    
    # Create and run worker
    worker = TaskWorker()
    try:
        worker.run()
    except KeyboardInterrupt:
        logger.info("Worker interrupted by user")
    except Exception as e:
        logger.error(f"Worker crashed: {e}")
        sys.exit(1)

File: workers/nodejs/worker.js
#!/usr/bin/env node
/**
 * Node.js Worker for Task Queue System
 * 
 * This worker polls SQS for tasks and processes them based on their type.
 * Supports multiple task types with extensible handler architecture.
 */

const AWS = require('aws-sdk');
const { Client } = require('pg');
const axios = require('axios');

// Configure AWS
AWS.config.update({
    region: process.env.AWS_REGION || 'us-east-1',
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
});

// Initialize SQS client
const sqs = new AWS.SQS();
const queueUrl = process.env.AWS_SQS_QUEUE_URL;

// Worker configuration
const workerId = `nodejs-worker-${process.pid}`;
let running = true;

// Initialize database client
const dbClient = new Client({
    connectionString: process.env.DATABASE_URL
});

// Task handlers mapping
const handlers = {
    email: handleEmailTask,
    data: handleDataTask,
    file: handleFileTask,
    api: handleApiTask,
    script: handleScriptTask,
    report: handleReportTask
};

// Logger utility
const logger = {
    info: (msg) => console.log(`[${new Date().toISOString()}] INFO: ${msg}`),
    error: (msg) => console.error(`[${new Date().toISOString()}] ERROR: ${msg}`)
};

// Graceful shutdown handler
process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);

async function shutdown() {
    logger.info(`Worker ${workerId} shutting down...`);
    running = false;
    await dbClient.end();
    process.exit(0);
}

// Main worker function
async function startWorker() {
    logger.info(`Worker ${workerId} starting...`);
    
    // Connect to database
    await dbClient.connect();
    logger.info('Connected to database');
    
    while (running) {
        try {
            // Poll SQS for messages
            const params = {
                QueueUrl: queueUrl,
                MaxNumberOfMessages: 1,
                WaitTimeSeconds: 20, // Long polling
                MessageAttributeNames: ['All']
            };
            
            const response = await sqs.receiveMessage(params).promise();
            const messages = response.Messages || [];
            
            for (const message of messages) {
                await processMessage(message);
            }
            
        } catch (error) {
            logger.error(`Error in worker loop: ${error.message}`);
            await sleep(5000); // Wait before retrying
        }
    }
}

// Process a single message
async function processMessage(message) {
    try {
        // Parse message body
        const body = JSON.parse(message.Body);
        const taskId = body.task_id;
        const taskType = body.type;
        const payload = body.payload || {};
        
        logger.info(`Processing task ${taskId} of type ${taskType}`);
        
        // Update task status to processing
        await updateTaskStatus(message.MessageId, 'processing');
        
        // Get handler for task type
        const handler = handlers[taskType];
        if (!handler) {
            throw new Error(`Unknown task type: ${taskType}`);
        }
        
        // Execute task handler
        const result = await handler(payload);
        
        // Mark task as completed
        await completeTask(message.MessageId, result);
        
        // Delete message from queue
        await sqs.deleteMessage({
            QueueUrl: queueUrl,
            ReceiptHandle: message.ReceiptHandle
        }).promise();
        
        logger.info(`Task ${taskId} completed successfully`);
        
    } catch (error) {
        logger.error(`Error processing message: ${error.message}`);
        await failTask(message.MessageId, error.message);
    }
}

// Database operations
async function updateTaskStatus(messageId, status) {
    const query = `
        UPDATE tasks 
        SET status = $1, worker_id = $2, started_at = CURRENT_TIMESTAMP
        WHERE message_id = $3
    `;
    await dbClient.query(query, [status, workerId, messageId]);
}

async function completeTask(messageId, result) {
    const query = `
        UPDATE tasks 
        SET status = 'completed', 
            result = $1,
            completed_at = CURRENT_TIMESTAMP
        WHERE message_id = $2
    `;
    await dbClient.query(query, [JSON.stringify(result), messageId]);
}

async function failTask(messageId, error) {
    const query = `
        UPDATE tasks 
        SET status = 'failed', 
            error_message = $1,
            completed_at = CURRENT_TIMESTAMP
        WHERE message_id = $2
    `;
    await dbClient.query(query, [error, messageId]);
}

// Task handlers
async function handleEmailTask(payload) {
    const { recipient, subject, body } = payload;
    
    // Simulate email sending
    logger.info(`Sending email to ${recipient}: ${subject}`);
    await sleep(2000);
    
    return {
        status: 'sent',
        recipient: recipient,
        timestamp: Date.now()
    };
}

async function handleDataTask(payload) {
    const { operation, data = [] } = payload;
    
    logger.info(`Processing data operation: ${operation}`);
    
    let result;
    switch (operation) {
        case 'sum':
            result = data.reduce((a, b) => a + b, 0);
            break;
        case 'average':
            result = data.length > 0 ? data.reduce((a, b) => a + b, 0) / data.length : 0;
            break;
        case 'count':
            result = data.length;
            break;
        default:
            result = data;
    }
    
    await sleep(1000);
    
    return {
        operation: operation,
        result: result,
        items_processed: data.length
    };
}

async function handleFileTask(payload) {
    const { operation, file_path } = payload;
    
    logger.info(`Processing file operation: ${operation} on ${file_path}`);
    
    // Simulate file operations
    await sleep(3000);
    
    return {
        operation: operation,
        file_path: file_path,
        status: 'completed',
        size_bytes: 1024 // Mock value
    };
}

async function handleApiTask(payload) {
    const { url, method = 'GET', headers = {}, data } = payload;
    
    logger.info(`Making API call: ${method} ${url}`);
    
    try {
        const response = await axios({
            method: method,
            url: url,
            headers: headers,
            data: data,
            timeout: 30000
        });
        
        return {
            status_code: response.status,
            response: response.data,
            headers: response.headers
        };
    } catch (error) {
        return {
            error: error.message,
            status: 'failed'
        };
    }
}

async function handleScriptTask(payload) {
    const { script_name, args = [] } = payload;
    
    logger.info(`Executing script: ${script_name} with args: ${args}`);
    
    // Simulate script execution
    await sleep(5000);
    
    return {
        script: script_name,
        args: args,
        exit_code: 0,
        output: 'Script executed successfully'
    };
}

async function handleReportTask(payload) {
    const { report_type, parameters = {} } = payload;
    
    logger.info(`Generating report: ${report_type}`);
    
    // Simulate report generation
    await sleep(10000);
    
    return {
        report_type: report_type,
        status: 'generated',
        file_url: `/reports/${report_type}_${Date.now()}.pdf`,
        pages: 42
    };
}

// Utility function
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Check required environment variables
const requiredEnv = [
    'AWS_REGION',
    'AWS_ACCESS_KEY_ID',
    'AWS_SECRET_ACCESS_KEY',
    'AWS_SQS_QUEUE_URL',
    'DATABASE_URL'
];

const missing = requiredEnv.filter(varName => !process.env[varName]);
if (missing.length > 0) {
    logger.error(`Missing required environment variables: ${missing.join(', ')}`);
    process.exit(1);
}

// Start the worker
startWorker().catch(error => {
    logger.error(`Worker crashed: ${error.message}`);
    process.exit(1);
});